cnxdb/migrations/20171005213556_first-last.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+
+
+def up(cursor):
+    cursor.execute("""
+-- Create a function that always returns the first non-NULL item
+CREATE OR REPLACE FUNCTION first_agg ( anyelement, anyelement )
+RETURNS anyelement LANGUAGE SQL IMMUTABLE STRICT AS $$
+        SELECT $1;
+$$;
+
+-- And then wrap an aggregate around it
+CREATE AGGREGATE FIRST (
+        sfunc    = first_agg,
+        basetype = anyelement,
+        stype    = anyelement
+);
+
+-- Create a function that always returns the last non-NULL item
+CREATE OR REPLACE FUNCTION last_agg ( anyelement, anyelement )
+RETURNS anyelement LANGUAGE SQL IMMUTABLE STRICT AS $$
+        SELECT $2;
+$$;
+
+-- And then wrap an aggregate around it
+CREATE AGGREGATE LAST (
+        sfunc    = last_agg,
+        basetype = anyelement,
+        stype    = anyelement
+)""")
+
+
+def down(cursor):
+    cursor.execute("""
+    DROP FUNCTION first_agg(anyelement, anyelement)  CASCADE;
+    DROP FUNCTION last_agg(anyelement, anyelement)  CASCADE""")
```

Comment:
Probably need something like ``DROP AGGREGATE IF EXISTS ...``.

Reply:
It looks like the aggregates are dropped when rolling back the migration.

This is the output of `dbmigrator --context=cnx-db --config=/etc/cnx/archive/app.ini rollback`:

```diff
@@ -1832,44 +1806,20 @@
     if item not in prev_sys_path:
         new_sys_path.append(item)
         sys.path.remove(item)
 sys.path[:0] = new_sys_path
 $$;
 
 
 ALTER FUNCTION venv.activate_venv() OWNER TO cnxarchive;
 
 SET search_path = public, pg_catalog;
-
---
--- Name: first(anyelement); Type: AGGREGATE; Schema: public; Owner: cnxarchive
---
-
-CREATE AGGREGATE first(anyelement) (
-    SFUNC = first_agg,
-    STYPE = anyelement
-);
-
-
-ALTER AGGREGATE public.first(anyelement) OWNER TO cnxarchive;
-
---
--- Name: last(anyelement); Type: AGGREGATE; Schema: public; Owner: cnxarchive
---
-
-CREATE AGGREGATE last(anyelement) (
-    SFUNC = last_agg,
-    STYPE = anyelement
-);
-
-
-ALTER AGGREGATE public.last(anyelement) OWNER TO cnxarchive;
 
 --
 -- Name: list(text); Type: AGGREGATE; Schema: public; Owner: cnxarchive
 --
 
 CREATE AGGREGATE list(text) (
     SFUNC = comma_cat,
     STYPE = text,
     INITCOND = ''
 );

```

Reply:
Right, the CASCADE get them.